Created a Ball class with the following fields: color and number and …

noleynik29 opened the pull request:
…override the toString() method

create an enum with different colors
implemented method getRandomColor() in class ColorSupplier. This method should return a random color, based on the enum values (use the Random class for this) Then created a Lottery class with the getRandomBall() method, which will return the ball. This method should create a ball with a random color and a random number (the maximum possible number should be 100) and return this ball from the method.

To test the program, in the main() method, created three balls using the getRandomBall() method from the Lottery class. And printed information about them in the console.

noleynik29 committed:
Created a Ball class with the following fields: color and number and override the toString() method
create an enum with different colors
implemented method getRandomColor() in class ColorSupplier. This method should return a random color, based on the enum values (use the Random class for this)
Then created a Lottery class with the getRandomBall() method, which will return the ball. This method should create a ball with a random color and a random number (the maximum possible number should be 100) and return this ball from the method.

To test the program, in the main() method, created three balls using the getRandomBall() method from the Lottery class. And printed information about them in the console.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Application.java:
```diff
@@ -3,5 +3,12 @@
 public class Application {
     public static void main(String[] args) {
         // create three balls using class Lottery and print information about them in console
+        Ball firstBall = Lottery.getRandomBall();
```
This violates checklist item: "Don't use static methods in your solution". You should create an instance of Lottery and call its instance method instead of using a static method.

mateacademy-ai-mentor replied:
This violates checklist item: "Don't use static methods in your solution". `getRandomBall()` should be an instance method, and you should create and reuse a `Lottery` instance.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Application.java:
```diff
@@ -3,5 +3,12 @@
         // create three balls using class Lottery and print information about them in console
+        Ball firstBall = Lottery.getRandomBall();
+        Ball secondBall = Lottery.getRandomBall();
+        Ball thirdBall = Lottery.getRandomBall();
```
This violates checklist item: "Use for loop for creating several objects instead of duplicating code blocks." You should use a loop to create and print the balls, not repeat the code three times.

mateacademy-ai-mentor replied:
This violates checklist item: "All magic numbers in your code should be constants." The number of balls (3) should be extracted as a constant.

mateacademy-ai-mentor replied:
This violates checklist item: "Don't use static methods in your solution". `getRandomBall()` should be an instance method, and you should create and reuse a `Lottery` instance.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Ball.java:
```diff
@@ -0,0 +1,14 @@
+package core.basesyntax;
+
+public class Ball {
+    String color;
```
This violates checklist item: "Pay attention to access modifiers." The fields should be private, not package-private.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Ball.java:
```diff
@@ -0,0 +1,14 @@
+
+public class Ball {
+    String color;
+    int number;
```
This violates checklist item: "Pay attention to access modifiers." The fields should be private, not package-private.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Ball.java:
```diff
@@ -0,0 +1,14 @@
+public class Ball {
+    String color;
+    int number;
+
```
This violates the requirement to have a constructor to initialize the fields. The Ball class should have a constructor that takes color and number as parameters.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ColorSupplier.java:
```diff
@@ -1,7 +1,11 @@
 public class ColorSupplier {
-    public String getRandomColor() {
-        return null;
+    public static String getRandomColor() {
```
This violates checklist item: "Don't use static methods in your solution." The method should be an instance method, not static.

mateacademy-ai-mentor replied:
This violates checklist item: "Don't use static methods in your solution". `getRandomColor()` should be an instance method, not static.

mateacademy-ai-mentor replied:
This violates the checklist item: "Don't use static methods." getRandomColor() should be an instance method, not static.

mateacademy-ai-mentor replied:
This violates the requirement: getRandomColor() should return a Colors enum value, not a String.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ColorSupplier.java:
```diff
@@ -1,7 +1,11 @@
-    public String getRandomColor() {
-        return null;
+    public static String getRandomColor() {
+        Random random = new Random();
```
This violates checklist item: "Don't create several instances of the same class if you can use one." A new Random instance is created every time; instead, Random should be a field of the class.

mateacademy-ai-mentor replied:
This violates checklist item: "Think about which variables should be local in the method and which should be class-level". The Random instance should be a class-level field, not created on each method call.

mateacademy-ai-mentor replied:
This violates the checklist item: "Don't create several instances of the same class." A new Random instance is created on each call. Instead, Random should be a class-level field and reused.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ColorSupplier.java:
```diff
@@ -1,7 +1,11 @@
 
+import java.util.Random;
+
 public class ColorSupplier {
```
This violates checklist item: "Pay attention to access modifiers." The class and its method should have explicit access modifiers (e.g., public).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Random;
+
+public class Lottery {
+    public static Ball getRandomBall() {
```
This violates checklist item: "Don't use static methods in your solution." The method should be an instance method, not static.

mateacademy-ai-mentor replied:
This violates checklist item: "Don't use static methods in your solution". `getRandomBall()` should be an instance method, not static.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+
+public class Lottery {
+    public static Ball getRandomBall() {
+        Ball ball = new Ball();
```
This violates the requirement that Ball should be constructed with color and number via a constructor, and its fields should be private. Direct field access is not allowed.

mateacademy-ai-mentor replied:
This violates the requirement that Ball should be constructed with color and number via a constructor, not using setters after creation.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+public class Lottery {
+    public static Ball getRandomBall() {
+        Ball ball = new Ball();
+        Random rand = new Random();
```
This violates checklist item: "Don't create several instances of the same class if you can use one." A new Random instance is created every time; instead, Random should be a field of the class.

mateacademy-ai-mentor replied:
This violates checklist item: "Think about which variables should be local in the method and which should be class-level". The Random instance should be a class-level field, not created on each method call.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+        Ball ball = new Ball();
+        Random rand = new Random();
+
+        ball.color = ColorSupplier.getRandomColor();
```
This violates checklist item: "Don't use static methods in your solution." ColorSupplier should be used as an instance, not via a static method.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+        Random rand = new Random();
+
+        ball.color = ColorSupplier.getRandomColor();
+        ball.number = rand.nextInt(100);
```
This violates checklist item: "All magic numbers in your code should be constants." The number 100 should be extracted as a constant.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+        ball.color = ColorSupplier.getRandomColor();
+        ball.number = rand.nextInt(100);
+
+        return ball;
```
This violates checklist item: "Don't create temporary variables if you can return the value directly from the method." You can return the new Ball directly.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision:** REQUEST_CHANGES — several core requirements and checklist items are violated, which are critical blockers.

**Critical issues to fix (with guidance):**
1. Application.java
   - Line(s): where `Lottery.getRandomBall()` is called three times
   - Issue: Using a static method violates “Don’t use static methods.” Also, code is duplicated and the magic number `3` is hardcoded.
   - Why: Core requirement is to use instance methods, a loop for multiple objects, and constants for magic numbers.
   - How: Create a `private static final int BALLS_COUNT = 3;`, instantiate `Lottery lottery = new Lottery();`, and use a `for` loop to generate and print balls via `lottery.getRandomBall()`.

2. Ball.java
   - Lines: field declarations for `color` and `number`
   - Issue: Fields are package-private and there is no constructor.
   - Why: Encapsulation is required; Ball should be constructed with color and number.
   - How: Make fields `private`, add a `public Ball(Colors color, int number)` constructor, and provide getters if needed. Ensure `toString()` uses these fields (optionally `color.name()`).

3. ColorSupplier.java
   - Lines: class and `getRandomColor()` method
   - Issues: Method is `static`; creates a new `Random` on each call; missing explicit access modifiers.
   - Why: Must avoid static for this behavior and reuse a single `Random` instance; access modifiers required.
   - How: Make class `public`, method `public Colors getRandomColor()`, and add `private final Random random = new Random();` as a field; use it inside the instance method.

4. Lottery.java
   - Lines: class and `getRandomBall()` method
   - Issues: Method is `static`; creates `Ball` via no-arg and sets fields directly; creates new `Random` each call; uses `ColorSupplier` statically; magic number `100`; unnecessary temp variable; missing access modifiers.
   - Why: Must use instance methods, encapsulated Ball construction, reused dependencies, constants for magic numbers, and clean returns.
   - How: Make class `public` with fields `private final Random random = new Random();` and `private final ColorSupplier colorSupplier = new ColorSupplier();`. Add `private static final int MAX_NUMBER = 100;`. Change to `public Ball getRandomBall()` returning `new Ball(colorSupplier.getRandomColor(), random.nextInt(MAX_NUMBER));`.

**Minor improvements:**
- Ensure all classes/methods have explicit access modifiers.
- Consider using `enum.name()` if specific string format is required.

You’re close structurally—great use of an enum and `toString()`. Refactoring as above will align your solution with OOP principles and the checklist. Keep going—you’ve got this!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

noleynik29 committed:
Created a Ball class with the following fields: color and number and override the toString() method
create an enum with different colors
implemented method getRandomColor() in class ColorSupplier. This method should return a random color, based on the enum values (use the Random class for this)
Then created a Lottery class with the getRandomBall() method, which will return the ball. This method should create a ball with a random color and a random number (the maximum possible number should be 100) and return this ball from the method.

To test the program, in the main() method, created three balls using the getRandomBall() method from the Lottery class. And printed information about them in the console.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Application.java:
```diff
@@ -3,5 +3,12 @@
     public static void main(String[] args) {
         // create three balls using class Lottery and print information about them in console
+        Ball firstBall = Lottery.getRandomBall();
+        Ball secondBall = Lottery.getRandomBall();
```
This violates checklist item: "Don't use static methods in your solution". `getRandomBall()` should be an instance method, and you should create and reuse a `Lottery` instance.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Application.java:
```diff
@@ -3,5 +3,12 @@
+        Ball firstBall = Lottery.getRandomBall();
+        Ball secondBall = Lottery.getRandomBall();
+        Ball thirdBall = Lottery.getRandomBall();
+
+        System.out.println(firstBall);
```
This violates checklist item: "Use for loop for creating several objects of the same class" and "All magic numbers in your code should be constants". You should use a constant for the number of balls and a loop to create and print them, rather than repeating code.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Ball.java:
```diff
@@ -0,0 +1,22 @@
+package core.basesyntax;
+
+public class Ball {
+    private String color;
```
This violates the requirement that the color field should be of the Colors enum type, not String. The checklist requires using an enum for color and tying it to Ball's color field.

mateacademy-ai-mentor replied:
This violates the requirement: The color field must be of the enum type (e.g., Colors), not String. Update the field type and all usages accordingly.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Ball.java:
```diff
@@ -0,0 +1,22 @@
+public class Ball {
+    private String color;
+    private int number;
+
```
This violates the requirement to have a constructor that takes both color and number as parameters. The Ball class should have a public constructor like `public Ball(Colors color, int number)`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Ball.java:
```diff
@@ -0,0 +1,22 @@
+    @Override
+    public String toString() {
+        return "Ball{"
+                + "color='" + color + '\''
```
The Ball class should use the enum type for color and, in toString(), use color.name() if color is an enum. This is required by the checklist item: "Use name() for getting String representation of enum constants".

mateacademy-ai-mentor replied:
This violates the requirement: In toString(), when using enum, you should use color.name() to get the string representation of the enum constant, not just color as a String.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ColorSupplier.java:
```diff
@@ -1,7 +1,11 @@
+    public static String getRandomColor() {
+        Random random = new Random();
+        Colors color = Colors.values()[random.nextInt(Colors.values().length)];
+        return color.name();
```
This violates the requirement that getRandomColor() should return a Colors enum value, not a String. The Ball class should use the enum directly for its color field.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+        Ball ball = new Ball();
+        Random rand = new Random();
+
+        ball.setColor(ColorSupplier.getRandomColor());
```
This violates checklist item: "Don't use static methods in your solution". ColorSupplier should be used as an instance, not statically.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+        Random rand = new Random();
+
+        ball.setColor(ColorSupplier.getRandomColor());
+        ball.setNumber(rand.nextInt(100));
```
This violates checklist item: "All magic numbers in your code should be constants". The value 100 should be replaced by a constant (e.g., MAX_NUMBER).

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: REQUEST_CHANGES**

Your solution has solid basics (the `Colors` enum looks good and `toString()` is overridden), but several critical checklist requirements are not met, which block approval.

**Critical issues to fix**
1. Static method usage (violates “Don’t use static methods in your solution”):
   - Application.java: using `Lottery.getRandomBall()` statically. Create a single `Lottery` instance and reuse it.
   - ColorSupplier.java: `getRandomColor()` is `static`. Make it an instance method and reuse a single instance where needed.
   - Lottery.java: `getRandomBall()` is `static`. Make it an instance method and call it on a `Lottery` object.
2. Ball design and enum usage:
   - Ball.java: `color` is `String`. It must be of type `Colors` enum. Update field type and all usages.
   - Ball.java: Missing required constructor. Add `public Ball(Colors color, int number)` and construct balls via this constructor (not setters).
   - Ball.java: In `toString()`, when using enum, prefer `color.name()` to get the string representation.
3. Instance and resource management:
   - ColorSupplier.java and Lottery.java: Don’t create `new Random()` on each call. Keep a single `Random` as a class-level field.
   - Lottery.java: Use a reusable `ColorSupplier` instance (non-static field) instead of static access.
4. Magic numbers and looping:
   - Lottery.java: Replace `100` with a `private static final int MAX_NUMBER = 100;` and use it.
   - Application.java: Replace hardcoded `3` with a constant (e.g., `BALLS_COUNT`) and use a `for` loop to create/print balls instead of duplicating code.
5. Access modifiers:
   - Ensure classes and their methods/constructors have explicit access modifiers (e.g., `public` for classes and public APIs).

**Why these are critical**
- They violate explicit checklist items and core task requirements (enum usage, instance-based design, constructor-based object creation), and will likely cause tests to fail.

**Hints for fixing**
- Refactor `Ball` first (enum field + constructor), then adjust `ColorSupplier` to return `Colors` and hold a single `Random`. Next, make `Lottery` an instance class with a class-level `Random` and a `ColorSupplier` field, returning a `Ball` via the constructor. Finally, in `Application.main`, create one `Lottery` and loop `BALLS_COUNT` times to print results.

You’re close—once these refactors are in, your solution should meet all requirements. Keep going!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

noleynik29 committed:
Created a Ball class with the following fields: color and number and override the toString() method
create an enum with different colors
implemented method getRandomColor() in class ColorSupplier. This method should return a random color, based on the enum values (use the Random class for this)
Then created a Lottery class with the getRandomBall() method, which will return the ball. This method should create a ball with a random color and a random number (the maximum possible number should be 100) and return this ball from the method.

To test the program, in the main() method, created three balls using the getRandomBall() method from the Lottery class. And printed information about them in the console.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Ball.java:
```diff
@@ -0,0 +1,22 @@
+package core.basesyntax;
+
+public class Ball {
```
This violates the requirement: Ball must have a constructor Ball(Colors color, int number) and objects should be created using this constructor, not via setters.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+
+public class Lottery {
+    public Ball getRandomBall() {
+        Ball ball = new Ball();
```
This violates the requirement: Ball objects must be created using a constructor that takes a Colors enum and an int, not via setters and default constructor.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+        Ball ball = new Ball();
+        Random rand = new Random();
+
+        ball.setColor(ColorSupplier.getRandomColor());
```
This violates the checklist item: Don't use static methods. ColorSupplier.getRandomColor() should be called on an instance, not statically. Also, ColorSupplier should be a field, not created each call.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+public class Lottery {
+    public Ball getRandomBall() {
+        Ball ball = new Ball();
+        Random rand = new Random();
```
This violates the checklist item: Don't create several instances of the same class. A new Random instance is created on each call. Instead, Random should be a class-level field and reused.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+        Random rand = new Random();
+
+        ball.setColor(ColorSupplier.getRandomColor());
+        ball.setNumber(rand.nextInt(100));
```
This violates the checklist item: All magic numbers in your code should be constants. The value 100 should be replaced with a private static final int MAX_NUMBER = 100 constant.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Educational approval: Nice progress overall—the Application class and the Colors enum look clean and fully aligned with requirements. A few key improvements will strengthen your solution and match the checklist.

Critical fixes to address:
- Ball.java
  - Line X: `private String color;` should be `private Colors color;` to meet the enum requirement and ensure type safety.
  - Line X: Add constructor `public Ball(Colors color, int number)` and use it for object creation (avoid setters). This aligns with the requirement to construct via constructor, not via setters.
  - Line X (toString): When using enum, print `color.name()` instead of treating color as `String`.
- ColorSupplier.java
  - Line X: `getRandomColor()` should be an instance method (not static) and return `Colors`, not `String`.
  - Line X: Hold a single `Random` as a private field (e.g., `private final Random random = new Random();`) and reuse it—don’t create a new one each call.
  - Hint: Use `Colors.values()` and pick by index; return the enum directly.
- Lottery.java
  - Line X: Use the `Ball(Colors color, int number)` constructor; don’t rely on setters/default constructor.
  - Line X: Keep a single `ColorSupplier` instance as a field and call its instance `getRandomColor()` method.
  - Line X: Keep a single `Random` as a field; avoid creating a new one per call.
  - Line X: Replace magic number `100` with `private static final int MAX_NUMBER = 100`.

Minor improvements:
- Prefer direct returns where possible to simplify methods.

You’re on the right track—Application and Colors are solid. Updating the enum usage, constructors, and instance management will bring everything into full compliance. Keep going!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

noleynik29 committed:
Created a Ball class with the following fields: color and number and override the toString() method
create an enum with different colors
implemented method getRandomColor() in class ColorSupplier. This method should return a random color, based on the enum values (use the Random class for this)
Then created a Lottery class with the getRandomBall() method, which will return the ball. This method should create a ball with a random color and a random number (the maximum possible number should be 100) and return this ball from the method.

To test the program, in the main() method, created three balls using the getRandomBall() method from the Lottery class. And printed information about them in the console.
